Reject/error on base declarations that appear after field declarations

dwblaikie opened the pull request.

dwblaikie committed:
Reject/error on base declarations that appear after field declarations

dwblaikie committed:
Move the diagnostic handling a bit further up, near other diagnostics.

dwblaikie commented:
Open to feedback on the diagnostic phrasing, where the handling is being done (too early? too late?) etc.

jonmeow commented on toolchain/check/handle_class.cpp:
```diff
@@ -512,6 +512,15 @@ auto HandleParseNode(Context& context, Parse::BaseDeclId node_id) -> bool {
 
+  if (!context.struct_type_fields_stack().PeekArray().empty()) {
+    CARBON_DIAGNOSTIC(BaseDeclAfterFieldDecl, Error,
+                      "`{0}` declaration appears after field declaration(s)",
```
```suggestion
                      "`{0}` declaration must come before field declarations",
```

Suggesting different phrasing to avoid the `(s)`

jonmeow replied:
Alternately:

```suggestion
                      "`{0}` declaration must be the first field declaration",
```

dwblaikie replied:
Hmm, I don't really think of the base as a field - but perhaps other folks/carbon wants to/does?

I went with the first option, but with "appear before" rather than "come before" because that sounded better to me and we have "appear before" and no "come before" (though we also have one "come after" and no "appear after")

jonmeow commented on toolchain/check/handle_class.cpp:
```diff
@@ -512,6 +512,15 @@ auto HandleParseNode(Context& context, Parse::BaseDeclId node_id) -> bool {
     return true;
   }
 
+  if (!context.struct_type_fields_stack().PeekArray().empty()) {
```
```suggestion
  if (auto fields = context.struct_type_fields_stack().PeekArray(); !fields.empty()) {
```

(see note suggestion below)

jonmeow commented on toolchain/check/handle_class.cpp:
```diff
@@ -512,6 +512,15 @@ auto HandleParseNode(Context& context, Parse::BaseDeclId node_id) -> bool {
+    CARBON_DIAGNOSTIC(BaseDeclAfterFieldDecl, Error,
+                      "`{0}` declaration appears after field declaration(s)",
+                      Lex::TokenKind);
+    context.emitter().Emit(node_id, BaseDeclAfterFieldDecl,
+                           Lex::TokenKind::Base);
```
Does it work to add a note for the first field location? i.e., something like:

```suggestion
    CARBON_DIAGNOSTIC(FirstFieldDecl, Note,
                      "First field declared here");
    context.emitter().Build(node_id, BaseDeclAfterFieldDecl,
                           Lex::TokenKind::Base).Note(fields.front(), FirstFieldDecl).Emit();
```

(I think `FirstFieldDecl` is a bad name, but I'm struggling with better naming)

dwblaikie replied:
For now the StructTypeFields don't have source locations for this extra note, but once we track the FieldDecls separately in https://github.com/carbon-language/carbon-lang/pull/4514  I can revisit this to add the note. I'll add a TODO for this.

jonmeow reviewed (commented):
Well, you asked for comments on the message. :)

jonmeow commented on toolchain/check/handle_class.cpp:
```diff
@@ -512,6 +512,15 @@ auto HandleParseNode(Context& context, Parse::BaseDeclId node_id) -> bool {
+                      "`{0}` declaration appears after field declaration(s)",
+                      Lex::TokenKind);
+    context.emitter().Emit(node_id, BaseDeclAfterFieldDecl,
+                           Lex::TokenKind::Base);
```
Why do you pass in `Base` here, versus putting `base` directly in the message?

dwblaikie replied:
Inspired by some nearby code (`DiagnoseClassSpecificDeclRepeated` - which admittedly is actually parameterized, with one caller using `Lex::TokenKind::Base` and another using `Lex::TokenKind::Adapt`, whereas this case only has the one value). Smallest argument in favor of type safety/write-it-once - there's one bit of code for mapping this keyword to text, rather than including freeform text that is consistent today, but isn't checked in any way... 
But I certainly don't feel strongly. I see there are at least several cases where it's written literally/in the string, so I'll change to that.

dwblaikie committed:
Rephrase diagnostic

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
`must come before` -> `must appear before`

(sounded better to me and we have one prior diagnostic example of
`appear before` and none for `come before`)

(admittedly we also have one example of "come after" and none for
"appear after"... )

dwblaikie committed:
Add TODO for adding a note with an example field decl

jonmeow reviewed (approved):
LG
